splitter.go

```diff
@@ -0,0 +1,135 @@
+
+// NewSplitReader returns a SplitReader.
+func NewSplitReader(rd Reader) SplitReader {
+	return SplitReader{rd}
+}
+
+func (r SplitReader) ReadTag(name string, value interface{}) error {
+	v := reflect.ValueOf(value)
+	if v.Kind() != reflect.Ptr {
+		return fmt.Errorf("ReadTag expects a pointer type but got %v", v.Kind())
+	}
+
+	err := error(nil)
+	switch v.Elem().Kind() {
+	case reflect.Struct:
+		str := v.Elem()
+		for i := 0; i < str.NumField(); i++ {
+			if str.Type().Field(i).PkgPath != "" {
+				continue // Type is not exported, so skip it
+			}
+
+			// Generate the name of the struct's field and recurse
+			fieldName, _ := getNameOfField(str, i)
+			if fieldName == "" {
+				continue // Can't touch that
+			}
+			fieldName = name + "." + fieldName // add prefix
+			if !str.Field(i).CanAddr() {
+				err = fmt.Errorf("Cannot address %s", fieldName)
+				break
+			}
+			fieldPointer := str.Field(i).Addr().Interface()
+
+			err = r.ReadTag(fieldName, fieldPointer)
+			if err != nil {
+				break
+			}
+		}
+	default:
+		// Just try with the underlying type
+		err = r.Reader.ReadTag(name, value)
+	}
+
+	return err
+}
+
+// SplitWriter splits writes of structs and arrays into separate writes of their components.
+type SplitWriter struct {
+	Writer
+}
+
+var _ = Writer(SplitWriter{}) // Compiler makes sure this type is a Writer
+
+// NewSplitWriter returns a SplitWriter.
+func NewSplitWriter(wr Writer) SplitWriter {
+	return SplitWriter{wr}
+}
+
+func (sw SplitWriter) WriteTag(name string, value interface{}) error {
+	v := reflect.ValueOf(value)
+	if v.Kind() == reflect.Ptr {
+		v = v.Elem() // Naturally use what the pointer is pointing to (but only do so once)
+	}
+
+	err := error(nil)
+	switch v.Kind() {
+	case reflect.Struct:
+		str := v
+		for i := 0; i < str.NumField(); i++ {
+			if str.Type().Field(i).PkgPath != "" {
+				continue // Type is not exported, so skip it
+			}
+
+			// Generate the name of the struct's field and recurse
+			fieldName, omitempty := getNameOfField(str, i)
+			if fieldName == "" {
+				continue // Can't touch that
+			}
+			if omitempty && str.Field(i).IsZero() {
+				continue // Skip zero values
+			}
+			fieldName = name + "." + fieldName // add prefix
+			fieldPointer := str.Field(i).Interface()
+
+			err = sw.WriteTag(fieldName, fieldPointer)
+			if err != nil {
+				break
+			}
+		}
+	default:
+		// Just try with the underlying type
+		err = sw.Writer.WriteTag(name, v.Interface())
+	}
+
+	return err
+}
+
+func getNameOfField(str reflect.Value, i int) (string, bool) {
+	omitEmpty := false
+
+	field := str.Type().Field(i)
+	plctag := field.Tag.Get("plctag")
+	if plctag == "" {
+		return field.Name, omitEmpty
+	}
+	opts := strings.Split(plctag, ",")
+	switch opts[0] {
+	case "-":
+		return "", omitEmpty // Ignore this field
+	case "":
+		opts[0] = field.Name // Use the field name as the name
+	}
+
+	for _, opt := range opts[1:] {
+		if opt == "omitempty" {
+			omitEmpty = true
```

Comment:
Suggestion: break out of the loop once you've found an omitEmpty occurrence?

Reply:
Tradeoffs: adding a break is slightly more efficient if the array has 2+ elements (but that should never occur anyway). But if one day other options are added, the break is slightly more likely to lead to buggy code if someone leaves it in.